### PR TITLE
Add grant_type=password support to GitLab/OpenID identity providers

### PR DESCRIPTION
**Pull request by liggitt**
GitLab and some OIDC providers support resource owner password grants, which lets us authenticate from the CLI using those providers

Allows setting `challenge: true` for OIDC and GitLab providers

### CODE REVIEW

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
@ironcladlou PTAL

File: pkg/auth/oauth/external/handler.go [hunk 1]
```diff
@@ -78,6 +81,70 @@ func (h *Handler) AuthenticationRedirect(w http.ResponseWriter, req *http.Reques
...
+	if err != nil {
+		if oauthErr, ok := err.(*osincli.Error); ok && oauthErr.Id == "invalid_grant" {
+			// An invalid_grant error means the username/password was rejected
+			return nil, false, nil
+		}
+		glog.V(4).Infof("Error getting access token using resource owner password grant: %v", err)
+		return nil, false, err
+	}
+
+	glog.V(4).Infof("Got access data")
```
Comment (ironcladlou) [inline pkg/auth/oauth/external/handler.go]: Add more context or remove

File: pkg/auth/oauth/external/handler.go [hunk 2]
```diff
@@ -78,6 +81,70 @@ func (h *Handler) AuthenticationRedirect(w http.ResponseWriter, req *http.Reques
...
+
+	identity, ok, err := h.provider.GetUserIdentity(accessData)
+	if err != nil {
+		glog.V(4).Infof("Error getting userIdentityInfo info: %v", err)
+		return nil, false, err
+	}
+	if !ok {
+		glog.V(4).Infof("Could not get userIdentityInfo info from access token")
```
Comment (ironcladlou) [inline pkg/auth/oauth/external/handler.go]: Log the username?

**Comment by ironcladlou** [comment]
Just a couple of trivial comments, looks great.

**Comment by liggitt** [comment]
[merge]

**Pull request merged**